src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -656,70 +653,204 @@ IEnumerator IEnumerable.GetEnumerator()
             return GetEnumerator();
         }
 
-        public IEnumerable<KeyValuePair<string, ModelState>> FindKeysWithPrefix(string prefix)
+        public static bool StartsWithPrefix(string prefix, string key)
         {
             if (prefix == null)
             {
                 throw new ArgumentNullException(nameof(prefix));
             }
 
-            ModelState exactMatchValue;
-            if (_innerDictionary.TryGetValue(prefix, out exactMatchValue))
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (StringComparer.OrdinalIgnoreCase.Equals(key, prefix))
+            {
+                return true;
+            }
+
+            if (key.Length <= prefix.Length)
             {
-                yield return new KeyValuePair<string, ModelState>(prefix, exactMatchValue);
+                return false;
             }
 
-            foreach (var entry in _innerDictionary)
+            if (!key.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
             {
-                var key = entry.Key;
+                if (key.StartsWith("[", StringComparison.OrdinalIgnoreCase))
+                {
+                    var subKey = key.Substring(key.IndexOf('.') + 1);
 
-                if (key.Length <= prefix.Length)
+                    if (!subKey.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return false;
+                    }
+
+                    if (string.Equals(prefix, subKey, StringComparison.Ordinal))
+                    {
+                        return true;
+                    }
+
+                    key = subKey;
```

Comment:
This is a strange case that makes `"[index].prefix"`, `"[index].prefix.property"` and so on a match for `"prefix"`. Do we have the history on why?